### PR TITLE
Support Vxworks

### PR DESCRIPTION
**Pull request by BaoshanPang**
Define the C interface to VxWorks

### CODE REVIEW

**Comment by gnzlbg** [comment]
This should at least test that libc builds correctly for this target (e.g. check ci/builds.sh, you might want to add all the vxworks target triples to the "rust nightly" environment variable).

**Comment by BaoshanPang** [comment]
> This should at least test that libc builds correctly for this target (e.g. check ci/builds.sh, you might want to add all the vxworks target triples to the "rust nightly" environment variable).

@gnzlbg 
I have added all the vxworks target triples but I think they are not going to build pass as they are not supported in rust yet. The PR to support vxWorks in rust is still on going:
https://github.com/rust-lang/rust/pull/61946

**Comment by gnzlbg** [comment]
> I have added all the vxworks target triples but I think they are not going to build pass as they are not supported in rust yet

You should probably add the target triples first, then libcore support, then libc, then liballoc, and then libstd support. WIth libcore support, the target triples here would work.

**Comment by gnzlbg** [comment]
Also, is it possible to run VxWorks under KVM or qemu on Linux ? If so, we'll need docker containers for these targets that do just that (e.g. see `ci/docker` for examples).

**Comment by gnzlbg** [comment]
ping

**Comment by BaoshanPang** [comment]
> Also, is it possible to run VxWorks under KVM or qemu on Linux ? If so, we'll need docker containers for these targets that do just that (e.g. see `ci/docker` for examples).

No, VxWorks can't run under KVM or qemu.

**Comment by BaoshanPang** [comment]
> ping

Sorry for the late response.
The PR for rust-lang/rust#61946 has been approved, the code will be upstreamed soon. I will let you know when the code are upstreamed so we can run the build test. 
The target triples have been updated so I also need to update the file

**Comment by gnzlbg** [comment]
Congrats and awesome.

>  I will let you know when the code are upstreamed so we can run the build test.

Just update this PR with the build tests, and then we can merge it.

**Comment by BaoshanPang** [comment]
>>rustup target add armv7-wrs-vxworks --toolchain stable
error: toolchain 'stable-x86_64-unknown-linux-gnu' does not contain component 'rust-std' for target 'armv7-wrs-vxworks'

I guess the rust version used for testing is not new enough to have the code to support vxWorks targets.

**Comment by gnzlbg** [comment]
The version used for testing is the latest nightly version. If the target
isn’t there, then no rust version exists with the target.

**Comment by BaoshanPang** [comment]
> The version used for testing is the latest nightly version. If the target isn’t there, then no rust version exists with the target.

My code was merged yerterday night some point, I guess these targets will show up with the next build.

**Comment by gnzlbg** [comment]
> My code was merged yerterday night some point, I guess these targets will show up with the next build.

Yes that's normal, it might take one day till the next nightly is released. You can locally try to see if your target is there. If it isn't then maybe publishing the `rust-src` component hasn't been properly set up for the target, and we would need to check what happened.

**Comment by BaoshanPang** [comment]
>>error: internal compiler error: src/librustc_codegen_llvm/context.rs:192: data-layout for builtin `armv7-unknown-linux-gnueabihf` target, `e-m:e-p:32:32-i64:64-v128:64:128-a:0:32-n32-S64`, differs from LLVM default, `e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64`

The PR to fix this issue has been created yesterday, it is waiting to be merged:
https://github.com/rust-lang/rust/pull/62745

**Comment by BaoshanPang** [comment]
>> Downloaded rand v0.6.5
  Downloaded syn v0.15.39
Command exited with code -1073741819

This seems not a problem in my code.

**Comment by BaoshanPang** [comment]
>>rustup target add armv7-wrs-vxworks --toolchain stable
error: toolchain 'stable-x86_64-unknown-linux-gnu' does not contain component 'rust-std' for target 'armv7-wrs-vxworks'

This is expected result as to build the "rust-std" for target armv7-wrs-vxworks, vxworks's work envionment(the compiler used to build C code, the linker to link libraries and executables etc) has to be installed somethere.

**Comment by BaoshanPang** [comment]
@gnzlbg 

Can you give any suggestion what should I do next?

Thanks,
Baoshan

File: ci/build.sh [hunk 1]
```diff
@@ -93,6 +93,13 @@ x86_64-unknown-freebsd \
 x86_64-unknown-linux-gnu \
 x86_64-unknown-linux-musl \
 x86_64-unknown-netbsd \
+armv7-wrs-vxworks \
```
Comment (gnzlbg) [inline ci/build.sh]: These are in the incorrect list of targets. You need to put them in the `RUST_NIGHTLY_LINUX_TARGETS` variable.

Comment (BaoshanPang) [reply ci/build.sh]: Hi gnzlbg,

I have moved these vxworks targets to RUST_NIGHTLY_LINUX_TARGETS.

Thanks,
Baoshan

**Comment by gnzlbg** [comment]
@BaoshanPang so apparently that didn't work, can you try moving them to `RUST_LINUX_NO_CORE_TARGETS` ?

**Comment by BaoshanPang** [comment]
Hi gnzlbg,

Can you help to check what's problem? It doesn't look like caused by my code.

>LLVM ERROR: invalid symbol redefinition
error: Could not compile `core`.


Thanks,
Baoshan

**Comment by gnzlbg** [comment]
> Can you help to check what's problem? It doesn't look like caused by my code.

This is an LLVM bug introduced with the upgrade to LLVM 9 in rust-lang/rust. There is a tracking issue there (https://github.com/rust-lang/rust/issues/62932) that has been reproduced and now an LLVM bug has been filled (https://bugs.llvm.org/show_bug.cgi?id=42760). 

You can try commenting out the thumbv6m-none-eabi target in the build.sh script (adding a FIXME linking to the rust-lang/rust issue).

**Comment by gnzlbg** [comment]
I've disabled the failing build test on master temporarily. Closing and reopening the PR to restart the CI builds.

**Comment by BaoshanPang** [comment]
> I've disabled the failing build test on master temporarily. Closing and reopening the PR to restart the CI builds.

Thanks for help.

File: src/vxworks/mod.rs [hunk 2]
```diff
@@ -0,0 +1,2322 @@
...
+use core::mem::size_of;
+
+#[cfg_attr(feature = "extra_traits", derive(Debug))]
+pub enum DIR{}
+impl ::Copy for DIR {}
+impl ::Clone for DIR {
+    fn clone(&self) -> DIR { *self }
+}
+
+/*
```
Comment (gnzlbg) [inline src/vxworks/mod.rs]: I think its ok to just remove these then. No need to mention them.

File: src/vxworks/mod.rs [hunk 3]
```diff
@@ -0,0 +1,2322 @@
+//! Hacking together the definitions for VxWorks Bindings
```
Comment (gnzlbg) [inline src/vxworks/mod.rs]: Could you run `rustfmt` on this file ?

e.g. `rustfmt src/vxworks/mod.rs` ? 

We have temporarily disabled the formatting check on CI, but as `rustfmt` fixes some bugs we plan to enable it again.

File: src/vxworks/mod.rs [hunk 4]
```diff
@@ -0,0 +1,2322 @@
...
+
+    // resource.h
+    pub struct rlimit { /* Is this really needed? Questionable ... */
+                           pub rlim_cur : ::size_t,
+                           pub rlim_max : ::size_t,
+    }
+
+    // stat.h
+    pub struct stat {
+                         pub st_dev       : ::dev_t,
```
Comment (gnzlbg) [inline src/vxworks/mod.rs]: Something is off with the alignment here.

File: src/vxworks/mod.rs (hunk 2 above)
Comment (BaoshanPang) [reply src/vxworks/mod.rs]: Sure.

File: src/vxworks/mod.rs (hunk 3 above)
Comment (BaoshanPang) [reply src/vxworks/mod.rs]: Sure.

File: src/vxworks/mod.rs (hunk 4 above)
Comment (BaoshanPang) [reply src/vxworks/mod.rs]: Sure.

**Comment by BaoshanPang** [comment]
>  Test timed out

Did you increas the waiting time for the newly added 7 vxworks targets?

**Comment by gnzlbg** [comment]
There is nothing to change, bors has just been timing out a lot lately.

@bors: retry

**Comment by BaoshanPang** [comment]
>Total 21048 (delta 0), reused 2 (delta 0), pack-reused 21041
HEAD is at 22c0f59fa814176ffcb80b38ce3580d6c19658aa.
Hard resetting to c21a3cf9db182ec54e23c6bd51a8f12e63d2c585...
Failed to force reset to c21a3cf9db182ec54e23c6bd51a8f12e63d2c585: object not found!

Is there anything I can do to move this forward?

**Comment by gnzlbg** [comment]
One cirrus build job is failing because some crate is failing to build. You could try to investigate and fix that, I don't know if anybody is currently doing that.

**Comment by BaoshanPang** [comment]
Can I ask bors to retry?
I see the issue has been reported and fixed here:
https://github.com/rust-lang/rust/issues/63115

**Comment by BaoshanPang** [comment]
@gnzlbg 
Can we let bors to retry? I see the fix for https://github.com/rust-lang/rust/issues/63115 has been merged into master 7 hours ago.

**Comment by BaoshanPang** [comment]
>>+rustup target add x86_64-unknown-redox --toolchain nightly
3448error: component 'rust-std' for target 'x86_64-unknown-redox' is unavailable for download for channel 'nightly'
3449If you require these components, please install and use the latest successful build version,

It seems there is still some problem in the test system.

**Comment by gnzlbg** [comment]
Sadly, the repo is in the middle of a transition to Azure Pipelines, and until it is finished nothing can be merged anymore.

**Comment by BaoshanPang** [comment]
> Sadly, the repo is in the middle of a transition to Azure Pipelines, and until it is finished nothing can be merged anymore.

Thanks for lettting me know it.  Do we have a time when the transition could be done? Thanks.

**Comment by gnzlbg** [comment]
>Do we have a time when the transition could be done? Thanks.

I don't think so. As soon as #1376 is merged.

File: src/lib.rs [hunk 5]
```diff
@@ -114,6 +114,12 @@ cfg_if! {
 
         mod switch;
         pub use switch::*;
+    } else if #[cfg(target_os = "vxworks")] {
+        mod fixed_width_ints;
+        pub use fixed_width_ints::*;
```
Comment (gnzlbg) [inline src/lib.rs]: These two aren't necessary - should be removed.

File: src/vxworks/mod.rs [hunk 6]
```diff
@@ -0,0 +1,2319 @@
...
+        pub sa_family : sa_family_t,
+        pub sa_data   : [::c_char; 14],
+    }
+
+    // socket.h
+    pub struct sockaddr_storage {
+        pub ss_len     : ::c_uchar,
+        pub ss_family  : ::sa_family_t,
+        pub __ss_pad1  : [::c_char; _SS_PAD1SIZE],
+        pub __ss_align : ::int32_t,
```
Comment (gnzlbg) [inline src/vxworks/mod.rs]: Should be i32.

File: src/vxworks/mod.rs [hunk 7]
```diff
@@ -0,0 +1,2319 @@
...
+pub const SOCK_SEQPACKET: ::c_int = 5;
+pub const SOCK_PACKET: ::c_int = 10;
+pub const SO_DEBUG: ::c_int = 0x0001;
+pub const SO_REUSEADDR: ::c_int = 0x0004;
+pub const SO_KEEPALIVE: ::c_int = 0x0008;
+pub const SO_DONTROUTE: ::c_int = 0x0010;
+pub const SO_RCVLOWAT: ::c_int = 0x0012;
+
+pub const _SS_MAXSIZE: usize = 128;
+pub const _SS_ALIGNSIZE: usize = size_of::<::uint32_t>();
```
Comment (gnzlbg) [inline src/vxworks/mod.rs]: Should be u32.

File: src/lib.rs (hunk 5 above)
Comment (BaoshanPang) [reply src/lib.rs]: Is it optional? I see they are added for every target.

Comment (gnzlbg) [reply src/lib.rs]: It's ok to add it. The content is deprecated, so it isn't super useful to add them to new targets, but at least with them old code that uses them will "just work" for this new target, so I guess there is value in that.

**Comment by gnzlbg** [comment]
@bors: r+ 

let's merge this!

**Comment by BaoshanPang** [comment]
@gnzlbg  
Please help to let bors to retry.

Thanks,
Baoshan

**Comment by gnzlbg** [comment]
@BaoshanPang thank you for working on this! Sorry that it took so long to merge it! 

Does VxWorks provide any kind of emulator to test applications on, e.g., x86 linux ?

**Comment by BaoshanPang** [comment]
> @BaoshanPang thank you for working on this! Sorry that it took so long to merge it!
> 
> Does VxWorks provide any kind of emulator to test applications on, e.g., x86 linux ?

We are using "simics" to test rust: https://en.wikipedia.org/wiki/Simics

**Comment by n-salim** [comment]
@gnzlbg thanks for your help getting this initial VxWorks support into Rust libc.

> Does VxWorks provide any kind of emulator to test applications on, e.g., x86 linux ?

We don't (yet) have a straightforward way for people other than Wind River's VxWorks customers to build and run VxWorks applications (including Rust executables), but this is something we hope to address in future.

We'd certainly like to be able to provide a way for "anyone" to try out Rust under VxWorks. Stay tuned!

**Comment by gnzlbg** [comment]
@n-salim until such a method is available, my recommendation is for Wind River to run libc tests internally at least, and submit PRs that fix them (fill issues if you have questions). That should catch most serious bugs in the implementation of the FFI wrapper.

If you want to get started with that check out the `ci/azure.yml` or `.cirrus.yml` script to see how our CI services do it. You don't necessarily need to use Docker containers, but can, e.g., just run `TARGET=... ci/run.sh` directly just like the OSX or FreeBSD targets do.

**Comment by n-salim** [comment]
@gnzlbg thanks very much for this information. We'll look into running the libc tests on VxWorks.

**Pull request merged**